File: frontend/src/clients/projects.ts
/**
 * Projects API Client
 *
 * This module provides API client methods for managing projects.
 * Includes built-in caching functionality to minimize API calls.
 */

import { getAuthToken } from "./auth-helper"

// API configuration
const API_BASE_URL = import.meta.env.VITE_API_URL || "http://localhost:8000"
const API_PREFIX = "/api/v1"

// Cache configuration
const CACHE_EXPIRATION_HOURS = 10
const CACHE_KEY_PROJECTS = "projects_list"

/**
 * Interface for cached data wrapper
 */
interface CachedData<T> {
  data: T
  timestamp: number
}

/**
 * Response interface for a single project
 */
export interface Project {
  project_id: string
  project_name: string
  description: string | null
  company_id: string | null
  created_by: string
  created_at: string // ISO datetime string
  is_active: boolean
}

/**
 * Response interface for project list
 */
export interface ProjectListResponse {
  projects: Project[]
  total_count: number
}

/**
 * Request interface for creating a project (simple)
 */
export interface ProjectCreateRequest {
  project_name: string
  description?: string
  company_id?: string
}

/**
 * Response interface for project creation
 */
export interface ProjectCreateResponse {
  project_id: string
  message: string
}

/**
 * Segment data for project setup
 */
export interface SegmentData {
  segment_name: string
  prompts: string
}

/**
 * Request interface for complete project setup
 */
export interface ProjectSetupRequest {
  project_name: string
  project_description?: string
  brand_name: string
  segments: SegmentData[]
}

/**
 * Response interface for project setup
 */
export interface ProjectSetupResponse {
  project_id: string
  brand_id: string
  prompt_count: number
  message: string
}

/**
 * Request interface for updating a project
 */
export interface ProjectUpdateRequest {
  project_name?: string
  description?: string
  is_active?: boolean
}

/**
 * Response interface for project update
 */
export interface ProjectUpdateResponse {
  project_id: string
  message: string
}

/**
 * Segment detail from backend
 */
export interface SegmentDetail {
  prompt_id: string
  segment_name: string
  prompts: string
  is_active: boolean
}

/**
 * Response interface for project details including brand and prompts
 */
export interface ProjectDetailResponse {
  project_id: string
  project_name: string
  description: string | null
  company_id: string | null
  created_by: string
  created_at: string
  is_active: boolean
  brand_id: string | null
  brand_name: string | null
  segments: SegmentDetail[]
}

/**
 * Request interface for full project update
 */
export interface ProjectFullUpdateRequest {
  project_name: string
  project_description?: string
  is_active: boolean
  brand_name: string
  segments: SegmentData[]
}

/**
 * Standard API error response
 */
export interface ApiError {
  detail: string
}

/**
 * Error type for 409 Conflict responses (e.g. brand already exists)
 */
export interface ApiConflictError extends Error {
  isConflict: boolean
}

/**
 * Projects API Client Class
 */
class ProjectsAPI {
  private readonly baseUrl: string
  private readonly apiPrefix: string

  constructor(baseUrl: string = API_BASE_URL, apiPrefix: string = API_PREFIX) {
    this.baseUrl = baseUrl
    this.apiPrefix = apiPrefix
  }

  /**
   * Get authorization headers
   */
  private async getAuthHeaders(): Promise<HeadersInit> {
    const token = await getAuthToken()
    return {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    }
  }

  /**
   * Check if cached data is still valid
   */
  private isCacheValid(timestamp: number): boolean {
    const now = Date.now()
    const expirationMs = CACHE_EXPIRATION_HOURS * 60 * 60 * 1000
    return now - timestamp < expirationMs
  }

  /**
   * Retrieve cached data from localStorage
   */
  private getCachedData<T>(key: string): T | null {
    try {
      const cachedString = localStorage.getItem(key)
      if (!cachedString) return null

      const cached: CachedData<T> = JSON.parse(cachedString)
      if (this.isCacheValid(cached.timestamp)) {
        console.log(`[ProjectsAPI] Using cached data for key: ${key}`)
        return cached.data
      }

      localStorage.removeItem(key)
      return null
    } catch {
      localStorage.removeItem(key)
      return null
    }
  }

  /**
   * Store data in localStorage cache
   */
  private setCachedData<T>(key: string, data: T): void {
    try {
      const cached: CachedData<T> = {
        data,
        timestamp: Date.now(),
      }
      localStorage.setItem(key, JSON.stringify(cached))
      console.log(`[ProjectsAPI] Data cached for key: ${key}`)
    } catch (error) {
      console.error(`[ProjectsAPI] Error caching data: ${error}`)
    }
  }

  /**
   * Clear projects cache
   */
  public clearCache(): void {
    localStorage.removeItem(CACHE_KEY_PROJECTS)
    console.log("[ProjectsAPI] Cache cleared")
  }

  /**
   * Fetch list of projects for the current user
   *
   * @param forceRefresh - If true, bypasses cache
   * @returns ProjectListResponse with projects array
   */
  async getProjects(forceRefresh: boolean = false): Promise<ProjectListResponse> {
    // Check cache first
    if (!forceRefresh) {
      const cached = this.getCachedData<ProjectListResponse>(CACHE_KEY_PROJECTS)
      if (cached) {
        return cached
      }
    }

    console.log("[ProjectsAPI] Fetching projects from API...")

    const url = `${this.baseUrl}${this.apiPrefix}/projects`

    const response = await fetch(url, {
      method: "GET",
      headers: await this.getAuthHeaders(),
    })

    if (response.status === 401) {
      // In case the token was invalidated
      console.error("[ProjectsAPI] Authentication failed, clearing cache and trying again...")
      this.clearCache()
      // Retry once with fresh headers and force refresh
      return await this.getProjects(true)
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to fetch projects")
    }

    const data: ProjectListResponse = await response.json()

    // Cache the result
    this.setCachedData(CACHE_KEY_PROJECTS, data)

    console.log(`[ProjectsAPI] Fetched ${data.total_count} projects`)
    return data
  }

  /**
   * Create a new project (simple version)
   *
   * @param projectData - Project creation data
   * @returns ProjectCreateResponse with new project ID
   */
  async createProject(
    projectData: ProjectCreateRequest
  ): Promise<ProjectCreateResponse> {
    console.log("[ProjectsAPI] Creating new project...")

    const url = `${this.baseUrl}${this.apiPrefix}/projects`

    const response = await fetch(url, {
      method: "POST",
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(projectData),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to create project")
    }

    const data: ProjectCreateResponse = await response.json()

    // Clear cache so next fetch gets fresh data
    this.clearCache()

    console.log(`[ProjectsAPI] Project created: ${data.project_id}`)
    return data
  }

  /**
   * Complete project setup with brand and prompts
   *
   * This creates:
   * 1. A new project
   * 2. A brand (if not exists)
   * 3. Brand prompts for each segment
   *
   * @param setupData - Complete project setup data
   * @returns ProjectSetupResponse with project ID, brand ID, and prompt count
   */
  async setupProject(
    setupData: ProjectSetupRequest
  ): Promise<ProjectSetupResponse> {
    console.log("[ProjectsAPI] Setting up project with brand and prompts...")

    const url = `${this.baseUrl}${this.apiPrefix}/projects/setup`

    const response = await fetch(url, {
      method: "POST",
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(setupData),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (response.status === 409) {
      const error: ApiError = await response.json()
      const conflictError = new Error(error.detail || "Brand already exists in another project")
      ;(conflictError as ApiConflictError).isConflict = true
      throw conflictError
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to setup project")
    }

    const data: ProjectSetupResponse = await response.json()

    // Clear cache so next fetch gets fresh data
    this.clearCache()

    console.log(
      `[ProjectsAPI] Project setup completed: project=${data.project_id}, brand=${data.brand_id}, prompts=${data.prompt_count}`
    )
    return data
  }

  /**
   * Update an existing project
   *
   * @param projectId - The project ID to update
   * @param updateData - Fields to update
   * @returns ProjectUpdateResponse with success message
   */
  async updateProject(
    projectId: string,
    updateData: ProjectUpdateRequest
  ): Promise<ProjectUpdateResponse> {
    console.log(`[ProjectsAPI] Updating project ${projectId}...`)

    const url = `${this.baseUrl}${this.apiPrefix}/projects/${projectId}`

    const response = await fetch(url, {
      method: "PATCH",
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(updateData),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (response.status === 404) {
      throw new Error("Project not found or you don't have permission to update it")
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to update project")
    }

    const data: ProjectUpdateResponse = await response.json()

    // Clear cache so next fetch gets fresh data
    this.clearCache()

    console.log(`[ProjectsAPI] Project updated: ${data.project_id}`)
    return data
  }

  /**
   * Get project details including brand and prompts
   *
   * @param projectId - The project ID to fetch
   * @returns ProjectDetailResponse with full project details
   */
  async getProjectDetail(projectId: string): Promise<ProjectDetailResponse> {
    console.log(`[ProjectsAPI] Fetching project detail ${projectId}...`)

    const url = `${this.baseUrl}${this.apiPrefix}/projects/${projectId}`

    const response = await fetch(url, {
      method: "GET",
      headers: await this.getAuthHeaders(),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (response.status === 404) {
      throw new Error("Project not found or you don't have permission to view it")
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to fetch project details")
    }

    const data: ProjectDetailResponse = await response.json()

    console.log(
      `[ProjectsAPI] Project detail fetched: ${data.project_id}, brand=${data.brand_name}, segments=${data.segments.length}`
    )
    return data
  }

  /**
   * Full update of project including brand and prompts
   *
   * @param projectId - The project ID to update
   * @param updateData - Full project update data
   * @returns ProjectSetupResponse with project ID, brand ID, and prompt count
   */
  async updateProjectFull(
    projectId: string,
    updateData: ProjectFullUpdateRequest
  ): Promise<ProjectSetupResponse> {
    console.log(`[ProjectsAPI] Full update project ${projectId}...`)

    const url = `${this.baseUrl}${this.apiPrefix}/projects/${projectId}`

    const response = await fetch(url, {
      method: "PUT",
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(updateData),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (response.status === 404) {
      throw new Error("Project not found or you don't have permission to update it")
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to update project")
    }

    const data: ProjectSetupResponse = await response.json()

    // Clear cache so next fetch gets fresh data
    this.clearCache()

    console.log(
      `[ProjectsAPI] Project full update completed: project=${data.project_id}, brand=${data.brand_id}, prompts=${data.prompt_count}`
    )
    return data
  }

  /**
   * Delete a project and its associated brand prompts
   *
   * @param projectId - The project ID to delete
   * @returns ProjectUpdateResponse with success message
   */
  async deleteProject(projectId: string): Promise<ProjectUpdateResponse> {
    console.log(`[ProjectsAPI] Deleting project ${projectId}...`)

    const url = `${this.baseUrl}${this.apiPrefix}/projects/${projectId}`

    const response = await fetch(url, {
      method: "DELETE",
      headers: await this.getAuthHeaders(),
    })

    if (response.status === 401) {
      throw new Error("Unauthorized - Please log in again")
    }

    if (response.status === 404) {
      throw new Error("Project not found or you don't have permission to delete it")
    }

    if (!response.ok) {
      const error: ApiError = await response.json()
      throw new Error(error.detail || "Failed to delete project")
    }

    const data: ProjectUpdateResponse = await response.json()

    // Clear cache so next fetch gets fresh data
    this.clearCache()

    console.log(`[ProjectsAPI] Project deleted: ${data.project_id}`)
    return data
  }
}

// Export singleton instance
export const projectsAPI = new ProjectsAPI()

File: backend/app/models/projects.py
"""
Projects Pydantic schemas for API request/response models.

This module defines the data transfer objects (DTOs) used by the projects API endpoints.
"""

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional
from datetime import datetime


class ProjectResponse(BaseModel):
    """
    Response schema for a single project.

    Attributes:
        project_id: Unique identifier for the project
        project_name: Display name of the project
        description: Project description (optional)
        company_id: Associated company ID (optional)
        created_by: User ID who created the project
        created_at: Timestamp when the project was created
        is_active: Whether the project is currently active
    """
    project_id: str = Field(..., description="Unique identifier for the project")
    project_name: str = Field(..., description="Project name")
    description: Optional[str] = Field(None, description="Project description")
    company_id: Optional[str] = Field(None, description="Associated company ID")
    created_by: str = Field(..., description="User ID who created the project")
    created_at: datetime = Field(..., description="Creation timestamp")
    is_active: bool = Field(True, description="Whether the project is active")

    model_config = ConfigDict(from_attributes=True)


class ProjectListResponse(BaseModel):
    """
    Response schema for a list of projects.

    Attributes:
        projects: List of project objects
        total_count: Total number of projects
    """
    projects: list[ProjectResponse] = Field(default_factory=list, description="List of projects")
    total_count: int = Field(0, description="Total number of projects")

    model_config = ConfigDict(from_attributes=True)


class SegmentData(BaseModel):
    """
    Schema for a single segment with prompts.

    Attributes:
        segment_name: Name of the market segment
        prompts: AI search prompts for this segment
    """
    segment_name: str = Field(..., min_length=1, max_length=100, description="Segment name")
    prompts: str = Field(..., min_length=1, description="Prompts for AI search")

    model_config = ConfigDict(from_attributes=True)


class ProjectSetupRequest(BaseModel):
    """
    Request schema for complete project setup including brand and prompts.

    This request creates:
    1. A new project record
    2. A brand record (if not exists)
    3. Brand prompt records for each segment

    Attributes:
        project_name: Name of the project
        project_description: Optional project description
        brand_name: Name of the brand to monitor
        segments: List of segments with their prompts (1-3 segments)
    """
    project_name: str = Field(..., min_length=1, max_length=100, description="Project name")
    project_description: Optional[str] = Field(None, max_length=255, description="Project description")
    brand_name: str = Field(..., min_length=1, max_length=100, description="Brand name to monitor")
    segments: list[SegmentData] = Field(..., min_length=1, max_length=3, description="Segments with prompts")

    model_config = ConfigDict(from_attributes=True)


class ProjectSetupResponse(BaseModel):
    """
    Response schema after complete project setup.

    Attributes:
        project_id: The newly created project's ID
        brand_id: The brand ID (existing or newly created)
        prompt_count: Number of prompts created
        message: Success message
    """
    project_id: str = Field(..., description="Newly created project ID")
    brand_id: str = Field(..., description="Brand ID (new or existing)")
    prompt_count: int = Field(0, description="Number of prompts created")
    message: str = Field("Project setup completed successfully", description="Success message")

    model_config = ConfigDict(from_attributes=True)


class ProjectCreateRequest(BaseModel):
    """
    Request schema for creating a new project (simple version).

    Attributes:
        project_name: Name of the project (required)
        description: Project description (optional)
        company_id: Associated company ID (optional)
    """
    project_name: str = Field(..., min_length=1, max_length=100, description="Project name")
    description: Optional[str] = Field(None, max_length=500, description="Project description")
    company_id: Optional[str] = Field(None, description="Associated company ID")

    model_config = ConfigDict(from_attributes=True)


class ProjectCreateResponse(BaseModel):
    """
    Response schema after creating a new project.

    Attributes:
        project_id: The newly created project's ID
        message: Success message
    """
    project_id: str = Field(..., description="Newly created project ID")
    message: str = Field("Project created successfully", description="Success message")

    model_config = ConfigDict(from_attributes=True)


class ProjectUpdateRequest(BaseModel):
    """
    Request schema for updating an existing project.

    All fields are optional - only provided fields will be updated.

    Attributes:
        project_name: New project name (optional)
        description: New project description (optional)
        is_active: New active status (optional)
    """
    project_name: Optional[str] = Field(None, min_length=1, max_length=100, description="Project name")
    description: Optional[str] = Field(None, max_length=500, description="Project description")
    is_active: Optional[bool] = Field(None, description="Whether the project is active")

    model_config = ConfigDict(from_attributes=True)


class ProjectUpdateResponse(BaseModel):
    """
    Response schema after updating a project.

    Attributes:
        project_id: The updated project's ID
        message: Success message
    """
    project_id: str = Field(..., description="Updated project ID")
    message: str = Field("Project updated successfully", description="Success message")

    model_config = ConfigDict(from_attributes=True)


class SegmentDetail(BaseModel):
    """
    Schema for a segment with full details from BrandPromptTable.

    Attributes:
        prompt_id: Unique identifier for the prompt record
        segment_name: Name of the market segment
        prompts: AI search prompts for this segment
        is_active: Whether this prompt is active
    """
    prompt_id: str = Field(..., description="Prompt record ID")
    segment_name: str = Field(..., description="Segment name")
    prompts: str = Field(..., description="Prompts for AI search")
    is_active: bool = Field(True, description="Whether this prompt is active")

    model_config = ConfigDict(from_attributes=True)


class ProjectDetailResponse(BaseModel):
    """
    Response schema for project details including brand and prompts.

    Attributes:
        project_id: Unique identifier for the project
        project_name: Display name of the project
        description: Project description (optional)
        company_id: Associated company ID (optional)
        created_by: User ID who created the project
        created_at: Timestamp when the project was created
        is_active: Whether the project is currently active
        brand_id: The brand ID associated with this project
        brand_name: The brand name associated with this project
        segments: List of segments with their prompts
    """
    project_id: str = Field(..., description="Unique identifier for the project")
    project_name: str = Field(..., description="Project name")
    description: Optional[str] = Field(None, description="Project description")
    company_id: Optional[str] = Field(None, description="Associated company ID")
    created_by: str = Field(..., description="User ID who created the project")
    created_at: datetime = Field(..., description="Creation timestamp")
    is_active: bool = Field(True, description="Whether the project is active")
    brand_id: Optional[str] = Field(None, description="Brand ID")
    brand_name: Optional[str] = Field(None, description="Brand name")
    segments: list[SegmentDetail] = Field(default_factory=list, description="Segments with prompts")

    model_config = ConfigDict(from_attributes=True)


class ProjectFullUpdateRequest(BaseModel):
    """
    Request schema for full project update including brand and prompts.

    Attributes:
        project_name: Name of the project
        project_description: Optional project description
        is_active: Whether the project is active
        brand_name: Name of the brand to monitor
        segments: List of segments with their prompts (1-3 segments)
    """
    project_name: str = Field(..., min_length=1, max_length=100, description="Project name")
    project_description: Optional[str] = Field(None, max_length=255, description="Project description")
    is_active: bool = Field(True, description="Whether the project is active")
    brand_name: str = Field(..., min_length=1, max_length=100, description="Brand name to monitor")
    segments: list[SegmentData] = Field(..., min_length=1, max_length=3, description="Segments with prompts")

    model_config = ConfigDict(from_attributes=True)
